File: implementation_generated_by_AI/Chapter_12_Binary_Search_Tree/Problems/12-3 Average node depth in a randomly built binary search tree/12-3.py
import random
from typing import List

def quicksort(arr: List[int], left: int, right: int) -> None:
    """
    Sort the array using the quicksort algorithm.
    :param arr: the array to be sorted
    :param left: the left index of the array, inclusive
    :param right: the right index of the array, inclusive
    """
    if left >= right:
        return

    pivot = arr[left]
    i = left + 1
    j = right

    while i <= j:
        while i <= j and arr[i] < pivot:
            i += 1
        while i <= j and arr[j] >= pivot:
            j -= 1
        if i < j:
            arr[i], arr[j] = arr[j], arr[i]

    arr[left], arr[j] = arr[j], arr[left]

    quicksort(arr, left, j - 1)
    quicksort(arr, j + 1, right)


def main() -> None:
    min_value = 0
    max_value = 100

    for n in range(1, 1001):
        random_vector = random.choices(range(min_value, max_value + 1), k=n)
        quicksort(random_vector, 0, len(random_vector) - 1)

        if not all(random_vector[i] <= random_vector[i + 1] for i in range(len(random_vector) - 1)):
            print("Error: The array is not sorted.")
            exit(1)

    print("All tests passed.")

if __name__ == "__main__":
    main()

File: implementation_generated_by_AI/Chapter_12_Binary_Search_Tree/Problems/12-3 Average node depth in a randomly built binary search tree/12-3.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <random>

/*
Sort the array using quicksort algorithm.
:param arr: the array to be sorted
:param left: the left index of the array, inclusive
:param right: the right index of the array, inclusive
*/
void quicksort(std::vector<int>& arr, int left, int right) {
    if (left >= right) {
        return;
    }

    int pivot = arr[left];
    int i = left + 1;
    int j = right;

    while (i <= j) {
        while (i <= j && arr[i] < pivot) {
            i++;
        }
        while (i <= j && arr[j] >= pivot) {
            j--;
        }
        if (i < j) {
            std::swap(arr[i], arr[j]);
        }
    }

    std::swap(arr[left], arr[j]);

    quicksort(arr, left, j - 1);
    quicksort(arr, j + 1, right);
}


// Function to generate a random vector of integers
std::vector<int> generate_random_vector(int n, int min_value = 0, int max_value = 100) {
    // Initialize a random device and random engine
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(min_value, max_value);

    // Generate the random vector
    std::vector<int> random_vector(n);
    for (int i = 0; i < n; ++i) {
        random_vector[i] = dis(gen);
    }
    return random_vector;
}

int main() {
    int min_value = 0; // Minimum value for random numbers
    int max_value = 100; // Maximum value for random numbers

    for (int n = 1; n <= 1000; n++) {
        std::vector<int> random_vector = generate_random_vector(n, min_value, max_value);

        quicksort(random_vector, 0, random_vector.size() - 1);

        if(std::is_sorted(random_vector.begin(), random_vector.end()) == false) {
            std::cout << "Error: The array is not sorted." << std::endl;
            exit(1);
        }
    }

    std::cout << "All tests passed." << std::endl;
    return 0;
}
